java/core/src/java/org/apache/orc/impl/RecordReaderUtils.java

```diff
@@ -553,6 +555,44 @@ private static void readDiskRanges(FSDataInputStream file,
     }
   }
 
+  /**
+   * Read the list of ranges from the file by updating each range in the list
+   */
+  private static void readDiskRangesVectored(
+      FSDataInputStream fileInputStream,
+      BufferChunkList range,
+      boolean doForceDirect) throws IOException {
+    if (range == null) return;
+
+    IntFunction<ByteBuffer> allocate =
+        doForceDirect ? ByteBuffer::allocateDirect : ByteBuffer::allocate;
+
+    var fileRanges = new ArrayList<FileRange>();
+    BufferChunk cur = range.get();
+    while (cur != null) {
+      if (!cur.hasData()) {
```

Comment:
`cur.hasData()` seems to make the logic complex here and line 581. Shall we exclude that case?